File: PriorityQueue/python/min_heap_priority_queue.py
import typing

E = typing.TypeVar("E")


class Full(Exception):
    pass


class Empty(Exception):
    pass


class Node(typing.Generic[E]):

    def __init__(self, key: int, element: E, index: int, tree: typing.Optional["MinHeapBinaryTree"] = None) -> None:
        self._key = key
        self._element = element
        self._tree = tree
        self._index = index

    @property
    def element(self) -> E:
        return self._element

    @property
    def index(self) -> int:
        return self._index

    @property
    def tree(self) -> "MinHeapBinaryTree":
        return self._tree

    @tree.setter
    def tree(self, tree: "MinHeapBinaryTree") -> None:
        if self._tree:
            raise ValueError("Tree is not empty!")
        self._tree = tree

    @property
    def parent(self) -> typing.Union["Node[E]", None]:
        return self._tree.get_parent(self._index)

    @property
    def left(self) -> typing.Union["Node[E]", None]:
        try:
            return self._tree.get_left(self._index)
        except IndexError:
            return None

    @property
    def right(self) -> typing.Union["Node[E]", None]:
        try:
            return self._tree.get_right(self._index)
        except IndexError:
            return None

    def children(self) -> list["Node[E]"]:
        output = []
        if self.left:
            output.append(self.left)
            if self.right:
                output.append(self.right)
        return output

    def has_children(self) -> bool:
        return len(self.children()) > 0

    def __repr__(self) -> str:
        return f"{self.__class__.__name__}({self._element})"

    def __lt__(self, other: "Node[E]") -> bool:
        return self._key < other._key

    def __gt__(self, other: "Node[E]") -> bool:
        return self._key > other._key


class MinHeapBinaryTree(typing.Generic[E]):

    def __init__(self, maxsize: int, root: typing.Optional[Node[E]] = None) -> None:
        if maxsize < 1:
            raise ValueError("Maxsize should be greater than or equal to one!")
        if root is not None:
            root.tree = self
        self._maxsize = maxsize
        self._list: list[typing.Union[Node[E], None]] = [None] * maxsize
        self._list[0] = root
        self._size = 0 if root is None else 1

    @property
    def depth(self) -> int:
        return self.height(self.root)

    @property
    def size(self) -> int:
        return self._size

    @property
    def maxsize(self) -> int:
        return self._maxsize

    @property
    def empty(self) -> bool:
        return self.root is None

    @property
    def is_full(self) -> bool:
        return self._size >= self._maxsize

    @property
    def root(self) -> typing.Union[Node[E], None]:
        return self._list[0]

    @root.setter
    def root(self, node: Node[E]) -> None:
        self._list[0] = node
        self._size = 1

    def _validate(self, node: Node[E]) -> None:
        if node.tree != self:
            raise ValueError("Node does not belong to this tree!")

    def _check_size(self, n: int) -> None:
        if (self._size + n) > self._maxsize:
            raise Full

    def get_parent(self, index: int) -> typing.Union[Node[E], None]:
        if index == 0:
            return None
        return self._list[(index - 1) // 2]

    def add_left(self, index: int, key: int, element: E) -> Node[E]:
        self._check_size(1)
        i = (2 * index) + 1
        node = Node(key, element, i, self)
        self._list[i] = node
        self._size += 1
        return node

    def get_left(self, index: int) -> Node[E]:
        return self._list[(2 * index) + 1]

    def add_right(self, index: int, key: int, element: E) -> Node[E]:
        self._check_size(1)
        i = (2 * index) + 2
        node = Node(key, element, i, self)
        self._list[i] = node
        self._size += 1
        return node

    def get_right(self, index: int) -> Node[E]:
        return self._list[(2 * index) + 2]

    def is_root(self, node: Node[E]) -> bool:
        self._validate(node)
        return (node == self.root) and (self.root is not None)

    def _swap(self, parent: Node[E], child: Node[E]) -> None:
        """Swaps parent position with the child's"""
        pindex, cindex = parent.index, child.index
        self._list[pindex] = child
        self._list[cindex] = parent
        parent._index = cindex
        child._index = pindex

    def _up_heap(self, node: Node[E]) -> None:
        if node is None:
            return
        parent = node.parent
        if parent is None:
            return
        if parent < node:
            return
        self._swap(parent, node)
        self._up_heap(node)

    def _down_heap(self, node: Node[E]) -> None:
        left, right = node.left, node.right
        if left:
            child_to_swap = left
            if right:
                if right < left:
                    child_to_swap = right
            self._swap(node, child_to_swap)
            self._down_heap(node)

    def insert(self, key: int, element: E) -> Node[E]:
        self._check_size(1)

        if self.empty:
            node = Node(key, element, 0, self)
            self.root = node
            return node

        parent = self.get_parent(self._size)
        if not parent.left:
            node = self.add_left(parent.index, key, element)
        else:
            node = self.add_right(parent.index, key, element)

        self._up_heap(node)
        return node

    def pop(self) -> typing.Union[Node[E], None]:
        if self.empty:
            raise Empty
        root = self.root

        self._swap(root, self._list[self._size - 1])
        self._list[root.index] = None
        self._down_heap(self._list[0])
        self._size -= 1
        return root

    @staticmethod
    def is_leaf(node: Node[E]) -> bool:
        return not node.has_children()

    @staticmethod
    def height(node: Node[E]) -> int:
        if node is None:
            return 0
        elif not node.has_children():
            return 1
        return 1 + max(MinHeapBinaryTree.height(node.left), MinHeapBinaryTree.height(node.right))

    @staticmethod
    def get_layers(node: Node[E]) -> list[list[Node[E]]]:
        if node is None:
            return []
        elif MinHeapBinaryTree.is_leaf(node):
            return [[node]]
        output = [[node]]
        children = node.children()
        while len(children) > 0:
            output.append(children)
            children_temp = []
            for c in children:
                children_temp.extend(c.children())

            children = children_temp

        return output

    @staticmethod
    def visualize_left_to_right(node: typing.Optional[Node[E]], level: typing.Optional[int]) -> str:

        def _vis(n: Node[E], lvl: int) -> list[str]:
            output = []
            if n.right:
                output.extend(_vis(n.right, lvl + 2))
            output.append(f"""{((" " * 4) * lvl)}-> {n.element}""")
            if n.left:
                output.extend(_vis(n.left, lvl + 2))
            return output

        if node is None:
            return ""
        results = _vis(node, level)
        results.insert(0, "\n")
        results.append("\n")
        return "\n".join(results)

    @staticmethod
    def visualize_top_to_bottom(node: typing.Optional[Node[E]], depth: int, start: typing.Optional[int] = 0) -> str:

        def _create_index(input_: list[list[Node[E]]]) -> int:
            flattened = [n for lst in input_ for n in lst]
            for idx, n in enumerate(flattened):
                n._index = idx
            return len(flattened)

        def _vis(n: Node[E], d: int, s: int) -> None:
            val = f"{n.element}" if n is not None else " "
            col = d ** 2
            n_str = f"""{" " * col}{" " * s}{val}"""
            results[n._index] = n_str
            if n.left:
                _vis(n.left, d - 1, s)
                if n.right:
                    _vis(n.right, d - 1, col + s)

        layers = MinHeapBinaryTree.get_layers(node)
        results = [""] * _create_index(layers)
        _vis(node, depth, start)
        output = [""]
        for i in range(depth):
            n = len(layers[i])
            lst = []
            for _ in range(n):
                try:
                    val = results.pop(0)
                    split = val.split(" ")
                    div = len("".join(lst))
                    lst.append(" ".join(split[div:]))
                except IndexError:
                    break
            output.append("".join(lst))
        output.append("\n")
        return "\n\n".join(output)

    def __iter__(self) -> typing.Iterator[Node[E]]:
        return iter(self._list)

    def __str__(self) -> str:
        return self.visualize_top_to_bottom(self.root, self.depth)


class PriorityQueue(typing.Generic[E]):

    def __init__(self, maxsize: int) -> None:
        self._list: MinHeapBinaryTree[E] = MinHeapBinaryTree(maxsize)

    @property
    def size(self) -> int:
        return self._list.size

    @property
    def empty(self) -> bool:
        return self._list.empty

    def insert(self, key: int, element: E) -> Node[E]:
        return self._list.insert(key, element)

    def min(self) -> typing.Union[Node[E], None]:
        if self.empty:
            return None
        return self._list.root

    def remove_min(self) -> typing.Union[Node[E], None]:
        if self.empty:
            return None
        return self._list.pop()

    def __str__(self) -> str:
        return str(self._list)


if __name__ == "__main__":
    queue: PriorityQueue[int] = PriorityQueue(6)
    queue.insert(0, 324)
    queue.insert(1, 53)
    queue.insert(5, 23)
    queue.insert(4, 32)
    queue.insert(10, 2)
    queue.insert(3, 23234)
    print(queue.size)
    print(queue.min())
    print(queue)

File: PriorityQueue/java/src/MinHeapPriorityQueue.java
import java.util.List;
import java.util.Arrays;
import java.util.ArrayList;

public class MinHeapPriorityQueue<E> {

    public static void main(String[] args) {

    }

    public static class Full extends Exception {}

    public static class Empty extends Exception {}

    public static class Node<E> implements Comparable<Node<E>> {
        private final int key;
        private final E element;
        private int index;
        private MinHeapBinaryTree<E> tree;

        private Node(int key, E element, int index, MinHeapBinaryTree<E> tree) {
            this.key = key;
            this.element = element;
            this.index = index;
            this.tree = tree;
        }

        private Node(int key, E element, int index) {
            this(key, element, index, null);
        }

        public int getKey() {
            return key;
        }

        public E getElement() {
            return element;
        }

        public int getIndex() {
            return index;
        }

        private void setIndex(int index) {
            this.index = index;
        }

        public MinHeapBinaryTree<E> getTree() {
            return tree;
        }

        private void setTree(MinHeapBinaryTree<E> tree) {
            this.tree = tree;
        }

        public Node<E> getParent() {
            return tree.getParent(index);
        }

        public Node<E> getLeft() {
            try {
                return tree.getLeft(index);
            } catch (IndexOutOfBoundsException e) {
                return null;
            }
        }

        public Node<E> getRight() {
            try {
                return tree.getRight(index);
            } catch (IndexOutOfBoundsException e) {
                return null;
            }
        }

        public List<Node<E>> children() {
            List<Node<E>> output = new ArrayList<>();
            if (getLeft() != null) {
                output.add(getLeft());
                if (getRight() != null) {
                    output.add(getRight());
                }
            }
            return output;
        }

        public boolean hasChildren() {
            return !children().isEmpty();
        }

        @Override
        public int compareTo(Node<E> other) {
            return Integer.compare(key, other.key);
        }

        @Override
        public String toString() {
            return String.format("%s(%s)", getClass().getName(), element);
        }
    }

    public static class MinHeapBinaryTree<E> {
        private final int maxSize;
        private final Node<E>[] list;
        private int size;

        @SuppressWarnings("all")
        private MinHeapBinaryTree(int maxSize, Node<E> root) {
            if (maxSize < 1)
                throw new IllegalArgumentException("Max size should be greater than or equal to 1!");
            this.maxSize = maxSize;
            this.list = (Node<E>[]) new Object[maxSize];
            if (root != null) {
                root.setTree(this);
                this.size = 1;
            } else {
                this.size = 0;
            }
            list[0] = root;
        }

        private MinHeapBinaryTree(int maxSize) {
            this(maxSize, null);
        }

        public int depth() {
            return height(getRoot());
        }

        public int size() {
            return size;
        }

        public int getMaxSize() {
            return maxSize;
        }

        public boolean isEmpty() {
            return size < 1;
        }

        public boolean isFull() {
            return size >= maxSize;
        }

        public Node<E> getRoot() {
            return list[0];
        }

        public void setRoot(Node<E> node) {
            list[0] = node;
            this.size = 1;
        }

        private void validate(Node<E> node) {
            if (node.getTree() != this)
                throw new IllegalArgumentException("Node does not belong to this tree!");
        }

        @SuppressWarnings("all")
        private void checkSize(int n) throws Full {
            if ((size + n) > maxSize)
                throw new Full();
        }

        public Node<E> getParent(int index) {
            if (index == 0)
                return null;
            return list[(index - 1) / 2];
        }

        public Node<E> addLeft(int index, int key, E element) throws Full {
            checkSize(1);
            int i = (2 * index) + 1;
            Node<E> node = new Node<>(key, element, i, this);
            list[i] = node;
            size++;
            return node;
        }

        public Node<E> getLeft(int index) {
            return list[(2 * index) + 1];
        }

        public Node<E> addRight(int index, int key, E element) throws Full {
            checkSize(1);
            int i = (2 * index) + 2;
            Node<E> node = new Node<>(key, element, i, this);
            list[i] = node;
            size++;
            return node;
        }

        public Node<E> getRight(int index) {
            return list[(2 * index) + 2];
        }

        public boolean isRoot(Node<E> node) {
            validate(node);
            return (node == getRoot()) && (getRoot() != null);
        }

        private void swap(Node<E> parent, Node<E> child) {
            int pindex = parent.getIndex();
            int cindex = child.getIndex();
            list[pindex] = child;
            list[cindex] = parent;
            parent.index = cindex;
            child.index = pindex;
        }

        private void upHeap(Node<E> node) {
            if (node == null)
                return;
            Node<E> parent = node.getParent();
            if (parent == null)
                return;
            if (parent.compareTo(node) < 0)
                return;
            swap(parent, node);
            upHeap(node);
        }

        private void downHeap(Node<E> node) {
            Node<E> left = node.getLeft();
            Node<E> right = node.getRight();
            if (left != null) {
                Node<E> childToSwap = left;
                if ((right != null) && (right.compareTo(left) < 0)) {
                    childToSwap = right;
                }
                swap(node, childToSwap);
                downHeap(node);
            }
        }

        public Node<E> insert(int key, E element) throws Full {
            checkSize(1);
            if (isEmpty()) {
                Node<E> node = new Node<>(key, element, 0, this);
                setRoot(node);
                return node;
            }
            Node<E> parent = getParent(size);
            Node<E> node;
            if (parent.getLeft() != null) {
                node = addLeft(parent.getIndex(), key, element);
            } else {
                node = addRight(parent.getIndex(), key, element);
            }
            upHeap(node);
            return node;
        }

        public Node<E> pop() throws Empty {
            if (isEmpty())
                throw new Empty();
            Node<E> root = getRoot();
            swap(root, list[size - 1]);
            list[root.getIndex()] = null;
            downHeap(list[0]);
            size--;
            return root;
        }

        public void visualize(Node<E> node, int level) {
        if (isEmpty() || node == null)
            return;
        if (node.getRight() != null)
            visualize(node.getRight(), level + 2);
        System.out.println(" ".repeat(4).repeat(level) + "-> " + node.getElement());
        if (node.getLeft() != null)
            visualize(node.getLeft(), level + 2);
    }

        public void visualize() {
            visualize(getRoot(), 0);
        }

        public static <T> boolean isLeaf(Node<T> node) {
            return !node.hasChildren();
        }

        public static <T> int height(Node<T> node) {
            if (node == null)
                return 0;
            else if (isLeaf(node))
                return 1;
            return 1 + Math.max(height(node.getLeft()), height(node.getRight()));
        }

        public static <T> List<List<Node<T>>> getLayers(Node<T> node) {
            List<List<Node<T>>> output = new ArrayList<>();
            if (node == null) {
                return output;
            } else if (isLeaf(node)) {
                output.add(List.of(node));
                return output;
            }
            output.add(List.of(node));
            List<Node<T>> children = node.children();
            while (!children.isEmpty()) {
                output.add(children);
                List<Node<T>> childrenTemp = new ArrayList<>();
                for (Node<T> c : children) {
                    childrenTemp.addAll(c.children());
                }
                children = childrenTemp;
            }
            return output;
        }
    }
}
